include/circt/Dialect/FIRRTL/InstanceGraph.h

```diff
@@ -0,0 +1,148 @@
+//===- InstanceGraph.h - Instance graph -------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the FIRRTL InstanceGraph, which is similar to a CallGraph.
+//
+//===----------------------------------------------------------------------===//
+#ifndef CIRCT_DIALECT_FIRRTL_INSTANCEGRAPH_H
+#define CIRCT_DIALECT_FIRRTL_INSTANCEGRAPH_H
+
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/SmallVector.h"
+#include <memory>
+
+namespace circt {
+namespace firrtl {
+
+class InstanceGraphNode {
+  using EdgeSetT = llvm::SmallVector<InstanceGraphNode *, 1>;
+
+public:
+  InstanceGraphNode(Operation *module) : module(module) {}
+  InstanceGraphNode(InstanceGraphNode &&other)
+      : module(other.module), instances(std::move(other.instances)) {
+    other.module = nullptr;
+  }
+
+  /// Get the module that this node is tracking.
+  Operation *getModule() const { return module; }
+
+  using iterator = EdgeSetT::iterator;
+  iterator begin() { return instances.begin(); }
+  iterator end() { return instances.end(); }
+
+  using const_iterator = EdgeSetT::const_iterator;
+  const_iterator begin() const { return instances.begin(); }
+  const_iterator end() const { return instances.end(); }
+
+private:
+  /// Record that a module instantiates this module.
+  void addInstance(InstanceGraphNode *node) { instances.push_back(node); }
+
+  /// The module.
+  Operation *module;
+
+  /// List of modules which instantiate this one.
+  EdgeSetT instances;
+
+  // Provide access to the constructor.
+  friend class InstanceGraph;
+};
+
+/// This graph tracks modules and where they are instantiated. This is the
+/// inverse of the regular instance graph.
+class InstanceGraph {
```

Comment:
I'm not sure how to parse:

> This is the inverse of the regular instance graph.

What is a regular instance graph?  Should this class be named "InverseInstanceGraph"?

Reply:
This code went through some iterations.. Looks like I need to clean it up.  I would say a regular graph is `module A instantiates module B`, while the inverse is `module B instantiated by module A`.

The graph right now is only tracking the information for inverse edges, but I tried to leave a big hole in the implementation so that the regular graph could be added.  If you look at the graph traits below, only the `llvm::Inverse` trait implemented.  This was done with the assumption that it would be faster to combine both graphs into 1 analysis, instead of building the graphs separately.

I can pretty easily add the forward edges + graph traits, it's just that I didn't need them for what I'm doing.  If there are no use-cases of the regular graph, then it would make a ton of sense to rename this to "InverseInstanceGraph".

I need to think a bit harder about what is needed here.

Reply:
Big update to how this is implemented: its now tracking all instance operations, and you can iterate the uses of a modules, as well as the instance ops in a module.  There are regular and inverse graph traits implemented.